lnrpc/websocket_proxy.go

```diff
@@ -0,0 +1,221 @@
+	p := &WebsocketProxy{
+		h:      h,
+		logger: logger,
+		upgrader: &websocket.Upgrader{
+			ReadBufferSize:  1024,
+			WriteBufferSize: 1024,
+			CheckOrigin: func(r *http.Request) bool {
+				return true
+			},
+		},
+	}
+	return p
+}
+
+// WebsocketProxy provides websocket transport upgrade to compatible endpoints.
+type WebsocketProxy struct {
+	h        http.Handler
+	logger   btclog.Logger
+	upgrader *websocket.Upgrader
+}
+
+// ServeHTTP handles the incoming HTTP request. If the request is an
+// "upgradeable" WebSocket request (identified by header fields), then the
+// WS proxy handles the request. Otherwise the request is passed directly to the
+// underlying REST proxy.
+func (p *WebsocketProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if !websocket.IsWebSocketUpgrade(r) {
+		p.h.ServeHTTP(w, r)
+		return
+	}
+	p.upgradeToWebSocketProxy(w, r)
+}
+
+// upgradeToWebSocketProxy upgrades the incoming request to a WebSocket, reads
+// one incoming message then streams all responses until either the client or
+// server quit the connection.
+func (p *WebsocketProxy) upgradeToWebSocketProxy(w http.ResponseWriter,
+	r *http.Request) {
+
+	var responseHeader http.Header
+	conn, err := p.upgrader.Upgrade(w, r, responseHeader)
+	if err != nil {
+		p.logger.Errorf("error upgrading websocket:", err)
+		return
+	}
+	defer func() {
+		err := conn.Close()
+		if err != nil {
+			p.logger.Errorf("WS: error closing upgraded conn: %v",
+				err)
+		}
+	}()
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+
+	requestBodyR, requestBodyW := io.Pipe()
+	request, err := http.NewRequestWithContext(
+		r.Context(), r.Method, r.URL.String(), requestBodyR,
+	)
+	if err != nil {
+		p.logger.Errorf("WS: error preparing request:", err)
+		return
+	}
+	for header := range r.Header {
+		headerName := textproto.CanonicalMIMEHeaderKey(header)
+		forward, ok := defaultHeadersToForward[headerName]
+		if ok && forward {
+			request.Header.Set(headerName, r.Header.Get(header))
+		}
+	}
+	if m := r.URL.Query().Get(MethodOverrideParam); m != "" {
+		request.Method = m
+	}
+
+	responseBodyR, responseBodyW := io.Pipe()
+	response := newInMemoryResponseWriter(responseBodyW)
+	go func() {
+		<-ctx.Done()
+		_ = requestBodyW.CloseWithError(io.EOF)
+		_ = responseBodyW.CloseWithError(io.EOF)
+		response.closed <- true
+	}()
+
+	go func() {
+		defer cancelFn()
+		p.h.ServeHTTP(response, request)
+	}()
+
+	// Read loop: Take messages from websocket and write to http request.
+	go func() {
+		defer cancelFn()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			_, payload, err := conn.ReadMessage()
+			if err != nil {
+				if isClosedConnError(err) {
+					p.logger.Tracef("WS: socket "+
+						"closed: %v", err)
+					return
+				}
+				p.logger.Errorf("error reading message: %v",
+					err)
+				return
+			}
+			_, err = requestBodyW.Write(append(payload, '\n'))
+			if err != nil {
```

Comment:
Right, this could be inefficient. Fixed it and also made the code a bit easier to understand in general by creating a dedicated `requestForwardingReader` and `responseForwardingWriter`.